pennylane/interfaces/torch.py

```diff
@@ -135,38 +170,47 @@ def forward(ctx, input_kwargs, *input_):
             return torch.from_numpy(res)
 
         @staticmethod
+        @once_differentiable
         def backward(ctx, grad_output):  # pragma: no cover
             """Implements the backwards pass QNode vector-Jacobian product"""
             # NOTE: This method is definitely tested by the `test_torch.py` test suite,
             # however does not show up in the coverage. This is likely due to
             # subtleties in the torch.autograd.FunctionMeta metaclass, specifically
             # the way in which the backward class is created on the fly
 
+            # determine the QNode variables which should be differentiated
+            diff_indices = None
+            non_diff_indices = set()
+
+            for arg, arg_variable in zip(ctx.saved_tensors, qnode.arg_vars):
+                if not getattr(arg, "requires_grad", False):
+                    indices = [i.idx for i in _flatten(arg_variable)]
+                    non_diff_indices.update(indices)
+
+            if non_diff_indices:
+                diff_indices = set(range(qnode.num_variables)) - non_diff_indices
+
             # evaluate the Jacobian matrix of the QNode
-            jacobian = qnode.jacobian(ctx.args, ctx.kwargs)
+            jacobian = qnode.jacobian(ctx.args, ctx.kwargs, wrt=diff_indices)
+            jacobian = torch.as_tensor(jacobian, dtype=grad_output.dtype)
 
-            if grad_output.is_cuda:  # pragma: no cover
-                grad_output_np = grad_output.cpu().detach().numpy()
-            else:
-                grad_output_np = grad_output.detach().numpy()
+            vjp = torch.transpose(grad_output.view(-1, 1), 0, 1) @ jacobian
+            vjp = vjp.flatten()
 
-            # perform the vector-Jacobian product
-            if not grad_output_np.shape:
-                temp = grad_output_np * jacobian
-            else:
-                temp = grad_output_np.T @ jacobian
+            if non_diff_indices:
+                # Torch requires we return a gradient of size (num_variables,)
+                res = torch.zeros([qnode.num_variables], dtype=grad_output.dtype)
+                indices = np.fromiter(diff_indices, dtype=np.int64)
+                res[indices] = vjp
+                vjp = res
 
             # restore the nested structure of the input args
-            temp = [
-                np.array(i) if not isinstance(i, np.ndarray) else i
-                for i in unflatten(temp.flat, ctx.args)
-            ]
-
-            # convert the result to torch tensors, matching
-            # the type of the input tensors
+            grad_input_list = [torch.as_tensor(i) for i in unflatten_torch(vjp, ctx.args)[0]]
```

Comment:
Isn't `vjp` already a torch tensor? And the unflattened version would be a list over torch tensors already? So maybe I'm missing the point of this line as opposed to just `unflatten_torch(vjp, ctx.args)[0]`

Reply:
It also seems similar to lines 212-213 but maybe I'm just confused :laughing:

Reply:
You're right, this is a subtlety I can't quite puzzle out.

Consider the following example:

```pycon
>>> model = [0.4, [0.1, 0.1], np.array([[1, 2], [3, 4]])]
>>> flat = torch.arange(7)
>>> unflatten_torch(flat, model)[0]
[tensor(0),
 [tensor(1), tensor(2)],
 [[tensor(3), tensor(4)], [tensor(5), tensor(6)]]]
```

However, what we actually want is a list of tensors:
```pycon
>>> torch.as_tensor(i) for i in unflatten_torch(flat, model)[0]]
[tensor(0), tensor([1, 2]), tensor([[3, 4], [5, 6]])]
```

`torch.as_tensor` converts a list of tensor elements into a single tensor, doing what we want.

Reply:
Ah, I solved it! I should be unflattening based on the saved input _tensors_, not the input tensors converted to NumPy arrays! I'm doing this correctly in the tensorflow interface, just not here.

Thanks Tom for catching this! I've changed this to 
```python
grad_input_list = unflatten_torch(vjp, ctx.saved_tensors)[0]
```

and it works perfectly 🙂

Reply:
Nice! :rocket: